cli/src/action/api/rbac.rs

```diff
@@ -0,0 +1,143 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+
+use reqwest::blocking::Client;
+use serde::Deserialize;
+
+use crate::CliError;
+
+const RBAC_PROTOCOL_VERSION: u32 = 1;
+
+#[derive(Debug, Deserialize)]
+pub struct Role {
+    pub role_id: String,
+    pub display_name: String,
+    pub permissions: Vec<String>,
+}
+
+#[derive(Deserialize)]
+pub struct RoleList {
+    #[serde(rename = "data")]
+    pub roles: VecDeque<Role>,
+    pub paging: Paging,
+}
+
+#[derive(Deserialize)]
+pub struct Paging {
+    next: String,
+    total: usize,
+    limit: usize,
+    offset: usize,
+}
+
+pub struct RoleIter<'a> {
+    url: &'a str,
+    auth: &'a str,
+    current_page: Option<Result<RoleList, CliError>>,
+    consumed: bool,
+}
+
+impl<'a> RoleIter<'a> {
+    pub fn new(base_url: &'a str, auth: &'a str) -> Self {
+        Self {
+            url: base_url,
+            auth,
+            current_page: Self::load_page(base_url, auth, "/authorization/roles"),
+            consumed: false,
+        }
+    }
+
+    fn load_page(base_url: &str, auth: &str, link: &str) -> Option<Result<RoleList, CliError>> {
+        let result = Client::new()
+            .get(&format!("{}{}", base_url, link))
+            .header("SplinterProtocolVersion", RBAC_PROTOCOL_VERSION)
+            .header("Authorization", auth)
+            .send()
+            .map_err(|err| CliError::ActionError(format!("Failed to fetch node ID: {}", err)))
+            .and_then(|res| {
+                let status = res.status();
+                if status.is_success() {
+                    res.json::<RoleList>().map_err(|_| {
+                        CliError::ActionError(
+                            "Request was successful, but received an invalid response".into(),
+                        )
+                    })
+                } else {
+                    let message = res
+                        .json::<super::ServerError>()
+                        .map_err(|_| {
+                            CliError::ActionError(format!(
+                                "List roles fetch request failed with status code '{}', but error \
+                                 response was not valid",
+                                status
+                            ))
+                        })?
+                        .message;
+
+                    Err(CliError::ActionError(format!(
+                        "Failed to fetch role page: {}",
+                        message
+                    )))
+                }
+            });
+
+        Some(result)
+    }
+}
+
+impl<'a> Iterator for RoleIter<'a> {
+    type Item = Result<Role, CliError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This function is hard to understand (at least for my morning brain). Is there a different way to write it that's easier to understand? If not, please add some more comments here and/or to the `current_page` field.

Reply:
Added extensive comments in the method.